src/tagstore.rs

```diff
@@ -0,0 +1,111 @@
+use serde::de::{Deserializer, Error};
+use serde::ser::{Serialize, Serializer};
+
+/// signals that a newtype is from a CBOR tag
+pub const CBOR_NEWTYPE_NAME: &str = "__cbor_tag";
+
+/// extensions for all serde serializers
+pub trait SerializerExt: Serializer {
+    /// basically serialize_newtype_struct with a cbor tag value
+    fn serialize_cbor_tagged<T: Serialize>(
+        self,
+        tag: u64,
+        value: &T,
+    ) -> core::result::Result<Self::Ok, Self::Error> {
+        tag_access::set_tag(Some(tag));
```

Comment:
One idea instead of having `set_tag` and `get_tag` would be that the serializer consumes the tag previously stored. This would avoid the need for an explicit `set_tag(None)` and assures that each tag is only used once. Is this possible?

Reply:
Yes. The reason I unset the tag is that I don't know that the serializer on "the other side of the fence" is a CBOR aware serializer. If it is not, the serializer would *not* consume the tag and it would remain set on this thread.

Then later on the same thread you could have a CBOR deserializer being called from a non CBOR format, and it would read the tag value.

This is far fetched, but possible. So setting the tag back to None seems safer.

Reply:
(note that if the [visit_newtype_struct](https://docs.serde.rs/serde/de/trait.Visitor.html#method.visit_newtype_struct) would take a name parameter, I would not have to set the tag to None and in fact it would not even have to be an option. I could just check if the name parameter in both serialisation and deserialisation and be 100% sure that the caller is serde_cbor and has just set the tag...

But unfortunately it does not...

Some kind of extension mechanism in serde itself sure would be nice, but it seems we won't get one any time soon...

Reply:
I wonder if serde would accept something like

```
Visitor:
visit_extra_data(kind: u64, data: &[u8])

Serializer:
serialize_extra_data(kind: u64, data: &[u8])
```

That would be very low level, but generic enough for a lot of use cases (json whitespace, tags of all sorts, ...). There would have to be some repo where people can reserve a kind id, but that would be outside of serde...